tiled/server/authentication.py

```diff
@@ -894,8 +894,29 @@ async def refresh_session(
     return new_tokens
 
 
-@base_authentication_router.delete("/session/revoke/{session_id}")
+@base_authentication_router.post("/session/revoke")
 async def revoke_session(
+    request: Request,
+    refresh_token: schemas.RefreshToken,
+    settings: BaseSettings = Depends(get_settings),
+    db=Depends(get_database_session),
+):
+    "Mark a Session as revoked so it cannot be refreshed again."
+    request.state.endpoint = "auth"
+    payload = decode_token(refresh_token.refresh_token, settings.secret_keys)
+    session_id = payload["sid"]
+    # Find this session in the database.
+    session = await lookup_valid_session(db, session_id)
+    if session is None:
+        raise HTTPException(409, detail=f"No session {session_id}")
+    session.revoked = True
+    db.add(session)
+    await db.commit()
+    return Response(status_code=204)
```

Comment:
Does creating an explicit `Response` object here imply that no cookie is being sent?  In the '/logout' route the cookie was explicitly deleted (expired) -- presumably because there it "intercepts" the response as a function parameter?

In any case, I think we want the browser to clear this cookie -- even though it has already been revoked by the server.

Reply:
Authentication can happen in two ways:
1. API key (shared secret), backed by an `APIKey` in the database, passed in a header, cookie, or query parameter
2. OAuth2 with JWT access and refresh tokens, backed by a `Session` in the database, passed in a header only

Clients use one or the other, not both. This new route is addressing (2), and no cookies are involved.

The now-deprecated `logout` route was targeting (1), but there aren't any clients that actually use it. The route was created somewhat speculatively in the early development stage.

Reply:
In OAuth2, I thought that once the refresh token is received by the requesting server, it should only be sent to the client via an http-only cookie. That way a long-lifetime refresh token is not available to client-side scripts or from other origins.  This is the cookie I was imagining should get deleted/expired for tidiness rather than security (because the session was already revoked) so that it does not interfere with subsequent login attempts or unauthenticated routes.

I suppose for the python client the http-only cookie has less meaning than for a typical web browser.

Reply:
Cookies make things easy, but they get complicated in this world of SPAs that can talk to one or more backend services. If the tiled server is a different domain from the SPA talking to it, then the the SPA [won't have access to the cookie.](https://auth0.com/docs/manage-users/cookies/spa-authenticate-with-cookies)

Perhaps not quite on track for this issue, but here's on approach to [refresh tokens and SPAs.](https://auth0.com/blog/securing-single-page-applications-with-refresh-token-rotation/)

Reply:
In OAuth2, the refresh and access tokens never go near cookies. They are placed in the response body of the initial authorization response, and in the response body of each refresh. The client then holds the refresh token in internal storage somehow (could be browser local storage, or in the case of the Python client `~/.cache/tiled/tokens/...` with restrictive permissions on that file). The client places the access token in the header of future requests, as in `Authorization: Bearer {access_token}`.

Reply:
You can follow that logic in the [Authentication Details](https://blueskyproject.io/tiled/reference/authentication.html) docs. You'll see the `access_token` and `refresh_token` in a response body. Notice that the only time `set-cookie` shows up is in relation to [CSRF](https://developer.mozilla.org/en-US/docs/Glossary/CSRF) protection, holding a random unique string that is not a token.